### PR TITLE
V1.3.0 feature statistics

### LINKED ISSUE #150: Add Student Statistics
As a tutor, I can view the participation and presence statistics of a student so that I don't have to calculate them myself.

### LINKED ISSUE #151: Add Session Statistics
As a tutor, I can view the participation and presence statistics of students in a session so that I don't have to calculate them myself.

### PR DESCRIPTION
**Pull request by YuunoKun**
Adds student and session statistics.

Closes #151 
Closes #150

### CODE REVIEW

File: src/main/java/atas/model/Model.java [hunk 1]
```diff
@@ -232,6 +232,22 @@
      */
     Index generateRandomStudentIndex();
 
+    /**
+<<<<<<< HEAD
```
Comment (erisjacey) [inline src/main/java/atas/model/Model.java]: Maybe this line could be removed?

File: src/main/java/atas/commons/core/statistics/StatisticalValues.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package atas.commons.core.statistics;
+
+public class StatisticalValues {
```
Comment (erisjacey) [inline src/main/java/atas/commons/core/statistics/StatisticalValues.java:6]: Perhaps some Javadocs could be added here?

File: src/main/java/atas/model/Model.java [hunk 3]
```diff
@@ -232,6 +232,22 @@
...
+     */
+    void refreshStatistics();
+
+    /**
+     * Recalculate the current student statistics.
+     */
+    void refreshStudentStatistics();
+
+    /**
+     * Recalculate the current session statistics.
```
Comment (erisjacey) [inline src/main/java/atas/model/Model.java]: Perhaps the starting word for each Javadoc comment could be "Recalculates"?

File: src/main/java/atas/model/session/ReadOnlySessionList.java [hunk 4]
```diff
@@ -23,7 +23,9 @@
 
     ReadOnlySessionList getCopy();
 
+    /**
+     * Update the current student list with the given list.
```
Comment (erisjacey) [inline src/main/java/atas/model/session/ReadOnlySessionList.java]: Perhaps the comment could start with "Updates"?

File: src/main/java/atas/model/session/SessionList.java [hunk 5]
```diff
@@ -222,6 +222,16 @@ public void updateStudentPresence(SessionName sessionName, IndexRange indexRange
         }
     }
 
+    /**
+     * recalculate the participation and presence statistics of the current sessions in the
```
Comment (erisjacey) [inline src/main/java/atas/model/session/SessionList.java]: Maybe this comment can start with "Recalculates"?

File: src/main/java/atas/model/statistics/SessionStatistics.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+     * If there is already a statistic of the same type, it will be replaced with the new one. Or else
+     * if there isn't a static of the same type, it will be added to the set.
+     */
+    public void replaceStatistics(Statistics statistics) {
+        stats.removeIf(statistics::isSameStats);
+        stats.add(statistics);
+    }
+
+    /**
+     * Replace statistics in the set with the given collection of statistics.
```
Comment (erisjacey) [inline src/main/java/atas/model/statistics/SessionStatistics.java]: Maybe each comment could start with "Replaces" instead?

File: src/main/java/atas/model/statistics/Statistics.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+public interface Statistics {
+
+    /**
+     * Return statistics data of students in the sesssion.
+     */
+    Statistics getSessionStatistics(Session session);
+
+    /**
+     * Return statistics data of a student in the student list given the session list.
+     */
```
Comment (erisjacey) [inline src/main/java/atas/model/statistics/Statistics.java]: Return -> Returns?

File: src/main/java/atas/model/statistics/Statistics.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    String getDataAsPercentage();
+
+    /**
+     * Determines if the statistics are the same
+     */
+    boolean isSameStats(Statistics statistics);
+
+    /**
+     * Obtain the statistical value of current statistics.
```
Comment (erisjacey) [inline src/main/java/atas/model/statistics/Statistics.java]: Obtain -> Obtains?

File: src/main/java/atas/model/statistics/StudentStatistics.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+     * If there is already a statistic of the same type, it will be replaced with the new one. Or else
+     * if there isn't a static of the same type, it will be added to the set.
+     */
+    public void replaceStatistics(Statistics statistics) {
+        stats.removeIf(statistics::isSameStats);
+        stats.add(statistics);
+    }
+
+    /**
+     * Replace statistics in the set with the given collection of statistics.
```
Comment (erisjacey) [inline src/main/java/atas/model/statistics/StudentStatistics.java]: Replace -> Replaces?

**Review by erisjacey — APPROVED** [review]
LGTM! Just a few nitpicks w.r.t. Javadoc comments.

File: src/main/java/atas/commons/core/statistics/StatisticalValues.java (hunk 2 above)
Comment (marcustw) [reply src/main/java/atas/commons/core/statistics/StatisticalValues.java:6]: Agreed with Eris.

File: src/main/java/atas/model/statistics/SessionStatistics.java [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+     */
+    public void addStatistics(Statistics... statistics) {
+        // do not change this to Collections.addAll, they are not the same.
+        for (Statistics stat: statistics) {
+            stats.add(stat);
+        }
+    }
+
+    /**
+     * Replace statistic in the set with the given one.
```
Comment (marcustw) [inline src/main/java/atas/model/statistics/SessionStatistics.java]: Perhaps could start with "Replaces" instead as mentioned by Eris.

File: src/main/java/atas/model/student/StudentList.java [hunk 11]
```diff
@@ -93,6 +97,19 @@ public void removeStudent(Student key) {
 
     //// util methods
 
+    /**
+     * Recalculate student statistics using the given session list.
```
Comment (marcustw) [inline src/main/java/atas/model/student/StudentList.java:101]: Perhaps start with "Recalculates" ?

File: src/main/java/atas/ui/statistics/StatisticsLabel.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
+package atas.ui.statistics;
+
+import static atas.commons.core.statistics.StatisticalValues.STAT_FULL;
+import static atas.commons.core.statistics.StatisticalValues.STAT_NULL;
+
+import atas.model.statistics.Statistics;
+import javafx.scene.control.Label;
+
+
```
Comment (marcustw) [inline src/main/java/atas/ui/statistics/StatisticsLabel.java:9]: Perhaps missing Javadocs for this class?

**Review by marcustw — APPROVED** [review]
LGTM

**Pull request merged**